pkg/controller/framework/shared_informer.go

```diff
@@ -0,0 +1,243 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/runtime"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+)
+
+// if you use this, there is one behavior change.  When you receive a notification, the cache will be AT LEAST as
+// current as the notification, but it MAY be more current.  So if there was a create, followed by a delete, the cache
+// would NOT have your item.  All controllers should already be deduping on their side (if you forget to brush your teeth
+// at bedtime, you don't do it twice in the morning), but I doubt many people have considered the problem
+// This has advantages over the broadcaster since it allows us to share a common cache across many controllers.
+// Extending the broadcaster would have required us keep duplicate caches for each watch.
+type SharedInformer interface {
+	AddEventHandler(handler ResourceEventHandler) error
```

Comment:
I think it's fine to assume that you can NOT unregister a handler - I don't see any usecase for it and it seems to simplfy things

Regarding the semantics of invoking them, I think that we should put an invariant that:
- handlers from a "single client" are delivered sequentially
- there are no assumptions between clients - for a given event it may be delivered in parallel to different clients.

WDYT?